apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/PartitionOptions.cs

```diff
@@ -0,0 +1,76 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and 
+// limitations under the License.
+
+using System.Threading;
+
+namespace Google.Cloud.Spanner.Data;
+
+/// <summary>
+/// An immutable class representing partiton options.
+/// The specified values of `PartitionSizeBytes`, `MaxPartitions` and `CancellationToken`
+/// are utilized in the RPC call while generating the partition tokens.
+/// The specified flag for `DataBoostEnabled` will be stored in the request property
+/// of the <see cref="CommandPartition"/> object and is used when executing the
+/// partition object. If `DataBoostEnabled` is set to true, the request will be
+/// executed using Spanner's independent compute resources.
+/// </summary>
+public sealed class PartitionOptions
+{
+    /// <summary>
+    /// Returns a default instance of <see cref="PartitionOptions"/> with databoost enabled flag set to false.
+    /// </summary>
+    public static PartitionOptions Default { get; } = new PartitionOptions(null, null, false);
+
+    /// <summary>
+    /// Returns a new instance with the same options as this one, but with databoost enabled flag set to true.
+    /// </summary>
+    /// <returns> A new instance of <see cref="PartitionOptions"/> class.</returns>
+    public PartitionOptions WithDataBoostEnabled => new(PartitionSizeBytes, MaxPartitions, true, CancellationToken);
+
+    internal long? PartitionSizeBytes { get; }
+
+    internal long? MaxPartitions { get; }
+
+    internal bool DataBoostEnabled { get; }
+
+    internal CancellationToken CancellationToken { get; }
+
+    /// <summary>
+    /// Creates a new instance of <see cref="PartitionOptions"/> class.
+    /// </summary>
+    /// <param name="partitionSizeBytes">
+    /// The desired data size for each partition generated.  The default for this
+    /// option is currently 1 GiB.  This is only a hint. The actual size of each
+    /// partition may be smaller or larger than this size request.
+    /// </param>
+    /// <param name="maxPartitions">
+    /// The desired maximum number of partitions to return.  For example, this may
+    /// be set to the number of workers available.  The default for this option
+    /// is currently 10,000. The maximum value is currently 200,000.  This is only
+    /// a hint.  The actual number of partitions returned may be smaller or larger than
+    /// this maximum count request.
+    /// </param>
+    /// <param name="dataBoostEnabled">
+    /// If set to true, the request will be executed using Spanner independent compute resources.
+    /// The default value for this option is false.
+    /// </param>
+    /// <param name="cancellationToken">An optional token for canceling the call.</param>
+    public PartitionOptions(long? partitionSizeBytes, long? maxPartitions, bool? dataBoostEnabled, CancellationToken cancellationToken = default)
```

Comment:
I'd prefer this to be internal. In our experience, public constructors end up being annoying when we want to add more parameters - particularly as one is currently optional here. (We made that mistake with CallSettings.)

Having a static property for the default, then "With" methods that return the new values, is a nicely extensible pattern.

(There's an exception if we're *really* certain that no additional parameters will be needed, but it feels like this is a case where there may be more in the future.)